### ISSUE TITLE
Type variables defined in typeshed's `builtins.pyi` are usable in all files

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#1483]
### Summary

We don't emit a diagnostic on this snippet... but we should, because `_T_co` is not defined 😆

```py
class SupportsNext:
    def __next__(self) -> _T_co:
        raise NotImplementedError
```

We infer `_T_co` as being available here because of its definition in `builtins.pyi` [here](https://github.com/python/typeshed/blob/07b69ab9e92dbe4ca55c48b1c23ebf34b41538f1/stdlib/builtins.pyi#L83). But that definition should be treated as private-to-the-stub, I think.

Possibly we should just filter out typevar definitions when falling back to the builtin scope in name lookup? The same could also be said for type-alias definitions and protocol definitions in `builtins.pyi`, but they're arguably both more useful and less confusing to have available (you might want to use them quoted in type annotations, e.g. `x: "_PositiveInteger"` wouldn't be an unreasonable type annotation.

**Comment by AlexWaygood** [comment]
Pyrefly, mypy and pyright all filter out all "type-check-only" symbols (private type aliases, private typevars/paramspecs/typevartuples, private protocols) from the builtins fallback:

```py
reveal_type(_PositiveInteger)  # mypy, pyright, pyrefly: "`_PositiveInteger` is undefined"
reveal_type(_T_co)  # mypy, pyright, pyrefly: "`_T_co` is undefined"
reveal_type(_SupportsSynchronousAnext)  # mypy, pyright, pyrefly: "`_T_co` is undefined"
```

Mypy and pyrefly allow you to use these types if you explicitly import them, however, whereas pyright just views them as entirely private to the stub:

```py
# pyright: ❌, mypy/pyrefly: ✅
from builtins import _PositiveInteger, _T_co, _SupportsSynchronousAnext
```

I'd prefer to go with mypy/pyrefly's behaviour here:
- I'd rather not treat the `builtins` module differently to any other `.pyi` module when it comes to explicitly imported symbols
- I think it's often useful to import typeshed's protocols/aliases in `if TYPE_CHECKING` blocks, so I wouldn't want to forbid users from ever importing these type-check-only symbols in `.py` files

**Comment by dhruvmanila** [comment]
This seems reasonable. I'm assuming that this would still take `__all__` into account like if the `_T_co` variable isn't included in the `__all__` of `module.pyi` then it would still raise a diagnostic.

**Comment by AlexWaygood** [comment]
> I'm assuming that this would still take `__all__` into account like if the `_T_co` variable isn't included in the `__all__` of `module.pyi` then it would still raise a diagnostic.

We don't take `__all__` into account currently -- we only take `__all__` into account for symbols that are defined via imports in stub files. (These are always assumed to be private-to-the-stub unless they are included in `__all__` or they use a redundant alias.)

We could add a rule that says that these type-check-only symbols are treated as completely undefined if `__all__` exists and they're not included in `__all__`. But again, I think that would end up being annoying if you have an explicit import of one of these symbols in an `if TYPE_CHECKING` block. And I'd prefer to keep our rules for `.py` files and `.pyi` files as consistent as possible -- we generally allow you to import something from a `.py` file, even if that symbol is not included in the `__all__` of that `.py` file.

**Comment by charliermarsh** [comment]
@AlexWaygood -- are you still planning to own this or do you want to un-assign?

**Comment by AlexWaygood** [comment]
Feel free to pick it up! IIRC, the direction I wanted to go in https://github.com/astral-sh/ruff/pull/21288 was to create a reusable `exists_at_runtime()` method that could be used from both `infer/builder.rs` and places like https://github.com/astral-sh/ruff/blob/e31dffc99317d5dde5773e3e8ed99a06e53b62e3/crates/ty_python_semantic/src/types/list_members.rs#L393-L421, so that we had a single source of truth for which symbols in stubs should be considered "type-check only". Shouldn't really be too tricky to implement, I don't think, I just never got round to picking it back up...

**Comment by charliermarsh** [comment]
On it!
